bundle/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/TcsController.scala

```diff
@@ -0,0 +1,203 @@
+package edu.gemini.seqexec.server
+
+/**
+ * Created by jluhrs on 7/30/15.
+ *
+ * Interface to change and retrieve the TCS state.
+ * Most of the code deals with representing the state of the TCS subsystems.
+ */
+trait TcsController {
+
+  import TcsController._
+
+  def getConfig: SeqAction[TcsConfig]
+
+  def guide(gc: GuideConfig): SeqAction[Unit]
+
+  def applyConfig(tc: TelescopeConfig, gtc: GuidersTrackingConfig, ge: GuidersEnabled, agc: AGConfig): SeqAction[Unit]
+}
+
+object TcsController {
+
+  case class Offset(p: Double, q: Double)
+
+  object Offset {
+
+    object Null extends Offset(0.0, 0.0)
+
+  }
+
+  type Wavelength = Double
+
+  sealed trait TipTiltSource
+
+  object TipTiltSource {
+
+    object PWFS1 extends TipTiltSource
+
+    object PWFS2 extends TipTiltSource
+
+    object OIWFS extends TipTiltSource
+
+    object GAOS extends TipTiltSource
+
+  }
+
+  sealed trait M1Source
+
+  object M1Source {
+
+    object PWFS1 extends M1Source
+
+    object PWFS2 extends M1Source
+
+    object OIWFS extends M1Source
+
+    object GAOS extends M1Source
+
+    object HRWFS extends M1Source
+
+  }
+
+  sealed trait M2GuideConfig
+
+  object M2GuideOff extends M2GuideConfig
+
+  case class M2GuideOn(coma: Boolean, source: Set[TipTiltSource]) extends M2GuideConfig
+
+  sealed trait M1GuideConfig
+
+  object M1GuideOff extends M1GuideConfig
+
+  case class M1GuideOn(coma: Boolean, source: Set[M1Source]) extends M2GuideConfig
+
+  sealed trait Beam
+
+  object Beam {
+
+    object A extends Beam
+
+    object B extends Beam
+
+    object C extends Beam
+
+  }
+
+  // Combined configuration of nod position (telescope orientation) and chop position (M2 orientation)
+  case class NodChop(nod: Beam, chop: Beam)
+
+  // TCS can be configured to update a guide probe position only for certain nod-chop positions.
+  sealed trait NodChopTrackingConfig {
+    def get(nodchop: NodChop): Boolean
+  }
+
+  // If x is of type ActiveNodChopTracking then ∃ a:NodChop ∍ x.get(a)
+  // How could I reflect that in the code?
+  sealed trait ActiveNodChopTracking extends NodChopTrackingConfig
+
+  object NodChopTrackingConfig {
+
+    object None extends NodChopTrackingConfig {
+      def get(nodchop: NodChop): Boolean = false
+    }
+
+    object Normal extends ActiveNodChopTracking {
+      def get(nodchop: NodChop): Boolean = nodchop.nod == nodchop.chop
+    }
+
+    // Must be a non-empty set
```

Comment:
Maybe you can add a `require(s.nonEmpty)`. I couldn't find an `NonEmptySet` like the one for `List`